{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bd566e24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "GenRev not found, steiner wont work\n",
      "Import error: Negex. Using keyword matching instead\n",
      "Import error: Funcassociate. Make sure that funcassociate is in toolbox!\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%reload_ext autoreload\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import repos_tools\n",
    "import os.path\n",
    "import dream\n",
    "import attila_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6ade1da3",
   "metadata": {},
   "outputs": [],
   "source": [
    "hgn_fpath = '/Users/jonesa7/CTNS/resources/hgnc/hgnc_complete_set.txt'\n",
    "ingmt_fpath = '../../results/2022-08-22-dream-modules/dream_consensus_modules-PPI-STRING_Consensus.gmt'\n",
    "outgmt_fpath = '../../results/2022-08-22-dream-modules/dream_consensus_modules-PPI-STRING_Consensus-entrez_id.gmt'\n",
    "if not os.path.exists(outgmt_fpath):\n",
    "    repos_tools.symbol2entrez_gmt(ingmt_fpath, outgmt_fpath, hgn_fpath)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "54ddf7c7",
   "metadata": {},
   "source": [
    "%%bash\n",
    "cd ../../results/2022-08-22-dream-modules\n",
    "PVALS=../../resources/GWAS/Bellenguez-2022/GCST90027158/harmonised/35379992-GCST90027158-MONDO_0004975-Build38-corrected.f.tsv.pval\n",
    "# run Pascal\n",
    "~/tools/PASCAL/Pascal --pval=$PVALS --runpathway=on --genesetfile=dream_consensus_modules-PPI-STRING_Consensus-entrez_id.gmt\n",
    "# check results\n",
    "head output/35379992-GCST90027158-MONDO_0004975-Build38-corrected.f.tsv.PathwaySet--dream_consensus_modules-PPI-STRING_Consensus-entrez_id--sum.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2d790ab7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chi2Pvalue</th>\n",
       "      <th>empPvalue</th>\n",
       "      <th>Genes</th>\n",
       "      <th>numGenes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod272</th>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.00074</td>\n",
       "      <td>[LAIR1, SLA, FCAR, FCGR3A, ITK, CD58, C7orf43,...</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod248</th>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.00094</td>\n",
       "      <td>[PIK3R1, SKAP2, SH3BP5, SHC4, PDGFB, DOK3, HCL...</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod191</th>\n",
       "      <td>0.000894</td>\n",
       "      <td>0.00115</td>\n",
       "      <td>[AVPR1B, GNG2, GNAQ, APP, SPIN3, CPE, PLCB3, P...</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod142</th>\n",
       "      <td>0.000907</td>\n",
       "      <td>0.00248</td>\n",
       "      <td>[HS3ST1, SDC1, B3GALT6, SDC3, SDC2, SLC35D2, H...</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod328</th>\n",
       "      <td>0.001421</td>\n",
       "      <td>0.00302</td>\n",
       "      <td>[MCOLN2, TMPRSS9, KATNAL2, ABI3, NKIRAS2, GPR1...</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod358</th>\n",
       "      <td>0.989155</td>\n",
       "      <td>0.98939</td>\n",
       "      <td>[CTIF, C1orf212, TSPAN31, SLC41A1, C8orf4, GPB...</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod47</th>\n",
       "      <td>0.992947</td>\n",
       "      <td>0.99270</td>\n",
       "      <td>[SPHK1, SPHK2, CERS3, CERK, SGPP2, CERS4, SGMS...</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod66</th>\n",
       "      <td>0.998098</td>\n",
       "      <td>0.99812</td>\n",
       "      <td>[ACACA, ACACB, SCP2D1, CPT1C, MS4A8B, HMGCL, A...</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod9</th>\n",
       "      <td>0.998261</td>\n",
       "      <td>0.99844</td>\n",
       "      <td>[OR1D2, OR5A1, OR4K15, OR51Q1, OR5AK2, OR5H15,...</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPI-STRING_Consensus_mod77</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[HIST2H2BE, HIST1H2AG, HIST1H2AB, HIST1H2AD, H...</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>377 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             chi2Pvalue  empPvalue  \\\n",
       "PPI-STRING_Consensus_mod272    0.000020    0.00074   \n",
       "PPI-STRING_Consensus_mod248    0.000110    0.00094   \n",
       "PPI-STRING_Consensus_mod191    0.000894    0.00115   \n",
       "PPI-STRING_Consensus_mod142    0.000907    0.00248   \n",
       "PPI-STRING_Consensus_mod328    0.001421    0.00302   \n",
       "...                                 ...        ...   \n",
       "PPI-STRING_Consensus_mod358    0.989155    0.98939   \n",
       "PPI-STRING_Consensus_mod47     0.992947    0.99270   \n",
       "PPI-STRING_Consensus_mod66     0.998098    0.99812   \n",
       "PPI-STRING_Consensus_mod9      0.998261    0.99844   \n",
       "PPI-STRING_Consensus_mod77          NaN        NaN   \n",
       "\n",
       "                                                                         Genes  \\\n",
       "PPI-STRING_Consensus_mod272  [LAIR1, SLA, FCAR, FCGR3A, ITK, CD58, C7orf43,...   \n",
       "PPI-STRING_Consensus_mod248  [PIK3R1, SKAP2, SH3BP5, SHC4, PDGFB, DOK3, HCL...   \n",
       "PPI-STRING_Consensus_mod191  [AVPR1B, GNG2, GNAQ, APP, SPIN3, CPE, PLCB3, P...   \n",
       "PPI-STRING_Consensus_mod142  [HS3ST1, SDC1, B3GALT6, SDC3, SDC2, SLC35D2, H...   \n",
       "PPI-STRING_Consensus_mod328  [MCOLN2, TMPRSS9, KATNAL2, ABI3, NKIRAS2, GPR1...   \n",
       "...                                                                        ...   \n",
       "PPI-STRING_Consensus_mod358  [CTIF, C1orf212, TSPAN31, SLC41A1, C8orf4, GPB...   \n",
       "PPI-STRING_Consensus_mod47   [SPHK1, SPHK2, CERS3, CERK, SGPP2, CERS4, SGMS...   \n",
       "PPI-STRING_Consensus_mod66   [ACACA, ACACB, SCP2D1, CPT1C, MS4A8B, HMGCL, A...   \n",
       "PPI-STRING_Consensus_mod9    [OR1D2, OR5A1, OR4K15, OR51Q1, OR5AK2, OR5H15,...   \n",
       "PPI-STRING_Consensus_mod77   [HIST2H2BE, HIST1H2AG, HIST1H2AB, HIST1H2AD, H...   \n",
       "\n",
       "                             numGenes  \n",
       "PPI-STRING_Consensus_mod272        67  \n",
       "PPI-STRING_Consensus_mod248        74  \n",
       "PPI-STRING_Consensus_mod191        31  \n",
       "PPI-STRING_Consensus_mod142        32  \n",
       "PPI-STRING_Consensus_mod328        55  \n",
       "...                               ...  \n",
       "PPI-STRING_Consensus_mod358        45  \n",
       "PPI-STRING_Consensus_mod47         28  \n",
       "PPI-STRING_Consensus_mod66         94  \n",
       "PPI-STRING_Consensus_mod9          66  \n",
       "PPI-STRING_Consensus_mod77         38  \n",
       "\n",
       "[377 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enrichment_fpath = '../../results/2022-08-22-dream-modules/output/35379992-GCST90027158-MONDO_0004975-Build38-corrected.f.tsv.PathwaySet--dream_consensus_modules-PPI-STRING_Consensus-entrez_id--sum.txt'\n",
    "enrichment = pd.read_csv(enrichment_fpath, sep='\\t', index_col='Name')\n",
    "\n",
    "# read gene sets\n",
    "with open(ingmt_fpath) as f:\n",
    "    D = dict()\n",
    "    for line in f:\n",
    "        lin = line.rstrip().split('\\t')\n",
    "        d = {lin[0]: lin[2:]}\n",
    "        D.update(d)\n",
    "gsets = pd.DataFrame(pd.Series(D), columns=['Genes'])\n",
    "\n",
    "# join enrichment and gene sets\n",
    "enrichment = pd.concat([enrichment, gsets], axis=1)\n",
    "enrichment['numGenes'] = enrichment.Genes.apply(lambda l: len(l))\n",
    "enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a7b3570f",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_pretty = enrichment.copy()\n",
    "enrichment_pretty['Genes'] = enrichment_pretty.Genes.apply(lambda l: ' '.join(l))\n",
    "enrichment_pretty.to_csv('../../results/2022-08-22-dream-modules/enrichment.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "77f78035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>numSnps</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Status</th>\n",
       "      <th>knowledge</th>\n",
       "      <th>TWAS</th>\n",
       "      <th>TWAS2+</th>\n",
       "      <th>agora</th>\n",
       "      <th>agora2+</th>\n",
       "      <th>DESudhir</th>\n",
       "      <th>AD DE APOE3-APOE3</th>\n",
       "      <th>AD DE APOE4-APOE4</th>\n",
       "      <th>APOE3-4 DE neuron</th>\n",
       "      <th>APOE3-4 DE astrocyte</th>\n",
       "      <th>APOE3-4 DE microglia</th>\n",
       "      <th>rMTA-APOE3-4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CYB561</th>\n",
       "      <td>chr17</td>\n",
       "      <td>61509664</td>\n",
       "      <td>61523722</td>\n",
       "      <td>-</td>\n",
       "      <td>1534</td>\n",
       "      <td>97</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_FAIL</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGSF23</th>\n",
       "      <td>chr19</td>\n",
       "      <td>45116939</td>\n",
       "      <td>45140081</td>\n",
       "      <td>+</td>\n",
       "      <td>147710</td>\n",
       "      <td>204</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_FAIL</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PVR</th>\n",
       "      <td>chr19</td>\n",
       "      <td>45147097</td>\n",
       "      <td>45169428</td>\n",
       "      <td>+</td>\n",
       "      <td>5817</td>\n",
       "      <td>239</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_FAIL</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLASRP</th>\n",
       "      <td>chr19</td>\n",
       "      <td>45542297</td>\n",
       "      <td>45574214</td>\n",
       "      <td>+</td>\n",
       "      <td>11129</td>\n",
       "      <td>247</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_FAIL</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RELB</th>\n",
       "      <td>chr19</td>\n",
       "      <td>45504706</td>\n",
       "      <td>45541456</td>\n",
       "      <td>+</td>\n",
       "      <td>5971</td>\n",
       "      <td>212</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_FAIL</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome     start       end strand  gene_id  numSnps        pvalue  \\\n",
       "CYB561      chr17  61509664  61523722      -     1534       97  1.000000e-12   \n",
       "IGSF23      chr19  45116939  45140081      +   147710      204  1.000000e-12   \n",
       "PVR         chr19  45147097  45169428      +     5817      239  1.000000e-12   \n",
       "CLASRP      chr19  45542297  45574214      +    11129      247  1.000000e-12   \n",
       "RELB        chr19  45504706  45541456      +     5971      212  1.000000e-12   \n",
       "\n",
       "                                      Status  knowledge  TWAS  TWAS2+  agora  \\\n",
       "CYB561  DAVIES_LOWPRECISION_FAREBROTHER_FAIL        0.0   0.0     0.0    0.0   \n",
       "IGSF23  DAVIES_LOWPRECISION_FAREBROTHER_FAIL        0.0   1.0     0.0    0.0   \n",
       "PVR     DAVIES_LOWPRECISION_FAREBROTHER_FAIL        0.0   1.0     1.0    0.0   \n",
       "CLASRP  DAVIES_LOWPRECISION_FAREBROTHER_FAIL        0.0   1.0     1.0    0.0   \n",
       "RELB    DAVIES_LOWPRECISION_FAREBROTHER_FAIL        0.0   0.0     0.0    0.0   \n",
       "\n",
       "        agora2+  DESudhir  AD DE APOE3-APOE3  AD DE APOE4-APOE4  \\\n",
       "CYB561      0.0       0.0                0.0                0.0   \n",
       "IGSF23      0.0       0.0                0.0                0.0   \n",
       "PVR         0.0       0.0                0.0                0.0   \n",
       "CLASRP      0.0       0.0                0.0                0.0   \n",
       "RELB        0.0       0.0                0.0                0.0   \n",
       "\n",
       "        APOE3-4 DE neuron  APOE3-4 DE astrocyte  APOE3-4 DE microglia  \\\n",
       "CYB561                0.0                   0.0                   0.0   \n",
       "IGSF23                0.0                   0.0                   0.0   \n",
       "PVR                   0.0                   0.0                   0.0   \n",
       "CLASRP                0.0                   0.0                   0.0   \n",
       "RELB                  0.0                   0.0                   0.0   \n",
       "\n",
       "        rMTA-APOE3-4  \n",
       "CYB561           0.0  \n",
       "IGSF23           0.0  \n",
       "PVR              0.0  \n",
       "CLASRP           0.0  \n",
       "RELB             0.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv('../../results/2022-08-11-pascal/results.csv', index_col=0).fillna(0)\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8893c161",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'NCK2 PLCG2 CD2AP INPP5D',\n",
       " '',\n",
       " 'PLCG2 ALK INPP5D BDNF CBL IRS1 GDNF',\n",
       " 'INPP5D',\n",
       " 'PIK3R5 ANGPT1 BDNF',\n",
       " 'DAPP1 CD22',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dream.module_memberships('PPI-STRING_Consensus_mod248', enrichment, genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f2ed8444",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['knowledge',\n",
       " 'TWAS',\n",
       " 'TWAS2+',\n",
       " 'agora',\n",
       " 'agora2+',\n",
       " 'DESudhir',\n",
       " 'AD DE APOE3-APOE3',\n",
       " 'AD DE APOE4-APOE4',\n",
       " 'APOE3-4 DE neuron',\n",
       " 'APOE3-4 DE astrocyte',\n",
       " 'APOE3-4 DE microglia',\n",
       " 'rMTA-APOE3-4']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gsets = genes.columns[genes.columns.to_list().index('knowledge'):].to_list()\n",
    "gsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f4e059d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"shell_port\": 55434,\n",
      "  \"iopub_port\": 55435,\n",
      "  \"stdin_port\": 55436,\n",
      "  \"control_port\": 55438,\n",
      "  \"hb_port\": 55437,\n",
      "  \"ip\": \"127.0.0.1\",\n",
      "  \"key\": \"f82e96e1-a5592aab54e9ae5cea48928b\",\n",
      "  \"transport\": \"tcp\",\n",
      "  \"signature_scheme\": \"hmac-sha256\",\n",
      "  \"kernel_name\": \"\"\n",
      "}\n",
      "\n",
      "Paste the above JSON into a file, and connect with:\n",
      "    $> jupyter <app> --existing <file>\n",
      "or, if you are local, you can connect with just:\n",
      "    $> jupyter <app> --existing kernel-5bf91ef1-9528-4a50-88bb-f98877a2b1e3.json\n",
      "or even just:\n",
      "    $> jupyter <app> --existing\n",
      "if this is the most recent Jupyter kernel you have started.\n"
     ]
    }
   ],
   "source": [
    "%connect_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f41b47a6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
